src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -11,22 +11,51 @@
 namespace Microsoft.CodeAnalysis.CSharp
 {
     /// <summary>
-    /// This portion of the binder converts deconstruction-assignment syntax (AssignmentExpressionSyntax nodes with the left being a tuple expression)
-    /// into a BoundDeconstructionAssignmentOperator (or bad node).
+    /// This portion of the binder converts deconstruction-assignment syntax (AssignmentExpressionSyntax nodes with the left
+    /// being a tuple expression or declaration expression) into a BoundDeconstructionAssignmentOperator (or bad node).
     /// </summary>
     internal partial class Binder
     {
-        private BoundExpression BindDeconstructionAssignment(AssignmentExpressionSyntax node, DiagnosticBag diagnostics)
+        private BoundExpression BindDeconstruction(AssignmentExpressionSyntax node, DiagnosticBag diagnostics)
         {
-            var left = (TupleExpressionSyntax)node.Left;
-            ArrayBuilder<DeconstructionVariable> checkedVariables = BindDeconstructionAssignmentVariables(left.Arguments, left, diagnostics);
+            var left = node.Left;
+            var right = node.Right;
 
+            if (node.IsDeconstructionDeclaration())
+            {
+                return BindDeconstructionDeclaration(node, left, right, diagnostics);
+            }
+
+            // We only parse assignment-only or declaration-only deconstructions at this point
+            AssertDeconstructionIsAssignment(left);
```

Comment:
The first check verifies that all the variables are declared. The second check verifies that all the variables are just assigned too.
The second check tells use that we are not in a mixed scenario (some assignments and some declarations), such as `(x, var y) = e;`, which only will be supported later.